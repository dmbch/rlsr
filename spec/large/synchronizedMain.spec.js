/* eslint-env node, jest */
const R = require('ramda');
const diff = require('recursive-diff').getDiff;
const transform = require('../../src/transform');
const getEnv = require('../fixtures/env-large.fixture');
const getMessage = require('../fixtures/message-large.fixture');

const applyMainScope = R.assoc('scope', 'four');
const applyMinorLevel = R.assoc('type', 'feat');
const applyOtherScope = R.assoc('scope', 'three');
const applyMultiScope = R.assoc('body', 'affects: two, three');
const applyMajorLevel = R.assoc('footer', 'BREAKING CHANGE: text ');

describe('large test: exact mode', () => {
  Date.now = jest.fn(() => 0);

  it('applies patch message', () => {
    // make the date function returd deterministic values

    const env = getEnv([applyMainScope(getMessage('foo'))]);
    env.config.mode = 'synchronizedMain';

    const exp = transform(env);

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies minor message', () => {
    const env = getEnv([applyMainScope(applyMinorLevel(getMessage('bar')))]);
    env.config.mode = 'synchronizedMain';
    const exp = transform(env);

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies major message and bumps all', () => {
    const env = getEnv([applyMainScope(applyMajorLevel(getMessage('baz')))]);
    env.config.mode = 'synchronizedMain';
    const exp = transform(env);

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies several messages with one breaking change', () => {
    const env = getEnv([
      applyMajorLevel(applyMainScope(getMessage('taz'))),
      getMessage('raz'),
      applyOtherScope(getMessage('roo'))
    ]);
    env.config.mode = 'synchronizedMain';

    const exp = transform(env);

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies messages with multiple affected packages', () => {
    const env = getEnv([
      applyMinorLevel(getMessage('waz')),
      applyMultiScope(getMessage('tar'))
    ]);
    env.config.mode = 'synchronizedMain';

    const exp = transform(env);

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies minor then major', () => {
    const env = getEnv([applyMinorLevel(getMessage('raz'))]);
    env.config.mode = 'synchronizedMain';

    const exp1 = transform(env);
    expect(diff(env, exp1)).toMatchSnapshot();

    exp1.messages = [applyOtherScope(applyMinorLevel(getMessage('rar')))];
    exp1.packages = R.values(exp1.packages);

    const exp2 = transform(exp1);
    expect(diff(exp1, exp2)).toMatchSnapshot();
  });

  it('applies major then minor', () => {
    const env = getEnv([applyMajorLevel(getMessage('raz'))]);
    env.config.mode = 'synchronizedMain';

    const exp1 = transform(env);
    expect(diff(env, exp1)).toMatchSnapshot();

    exp1.messages = [applyOtherScope(applyMinorLevel(getMessage('rar')))];
    exp1.packages = R.values(exp1.packages);

    const exp2 = transform(exp1);

    expect(diff(exp1, exp2)).toMatchSnapshot();
  });
});
