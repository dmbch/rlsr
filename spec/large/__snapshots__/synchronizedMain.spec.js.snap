// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`large test: exact mode applies major message and bumps all 1`] = `
Object {
  "/changelog/5.0.0": Object {
    "operation": "add",
    "value": Object {
      "changes": Object {
        "five": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "four",
              ],
              "subject": "dependency update: fix: baz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
        "four": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "subject": "baz",
              "type": "fix",
            },
          ],
          "version": "5.0.0",
        },
        "one": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "four",
              ],
              "subject": "dependency update: fix: baz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
        "three": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "four",
              ],
              "subject": "dependency update: fix: baz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
        "two": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "four",
              ],
              "subject": "dependency update: fix: baz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "/mainPackage/rlsr/determinedIncrementLevel": Object {
    "operation": "add",
    "value": 2,
  },
  "/mainPackage/rlsr/previouslyUnreleased": Object {
    "operation": "add",
    "value": Array [],
  },
  "/mainPackage/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/messages/0": Object {
    "operation": "delete",
  },
  "/packages/five/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/four/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/one/dependencies/five": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/one/dependencies/four": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/one/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/three/dependencies/four": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/three/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/two/dependencies/three": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/two/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
}
`;

exports[`large test: exact mode applies major then minor 1`] = `
Object {
  "/changelog/5.0.0": Object {
    "operation": "add",
    "value": Object {
      "changes": Object {
        "five": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "one",
              ],
              "subject": "dependency update: fix: raz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
        "four": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "one",
              ],
              "subject": "dependency update: fix: raz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
        "one": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "BREAKING CHANGE: text ",
              "subject": "raz",
              "type": "fix",
            },
          ],
          "version": "5.0.0",
        },
        "three": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "one",
              ],
              "subject": "dependency update: fix: raz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
        "two": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "one",
              ],
              "subject": "dependency update: fix: raz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "/mainPackage/rlsr/determinedIncrementLevel": Object {
    "operation": "add",
    "value": 2,
  },
  "/mainPackage/rlsr/previouslyUnreleased": Object {
    "operation": "add",
    "value": Array [],
  },
  "/mainPackage/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/messages/0": Object {
    "operation": "delete",
  },
  "/packages/five/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/four/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/one/dependencies/five": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/one/dependencies/four": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/one/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/three/dependencies/four": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/three/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/two/dependencies/three": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/two/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
}
`;

exports[`large test: exact mode applies major then minor 2`] = `
Object {
  "/mainPackage/rlsr/determinedIncrementLevel": Object {
    "operation": "update",
    "value": -1,
  },
  "/messages/0": Object {
    "operation": "delete",
  },
  "/packages": Object {
    "operation": "update",
    "value": Object {
      "0": Object {
        "dependencies": Object {
          "five": "^5.0.0",
          "four": "^5.0.0",
        },
        "name": "one",
        "version": "5.0.0",
      },
      "1": Object {
        "dependencies": Object {
          "three": "^5.0.0",
        },
        "name": "two",
        "version": "5.0.0",
      },
      "2": Object {
        "dependencies": Object {
          "four": "^5.0.0",
        },
        "name": "three",
        "version": "5.0.0",
      },
      "3": Object {
        "name": "four",
        "version": "5.0.0",
      },
      "4": Object {
        "name": "five",
        "version": "5.0.0",
      },
    },
  },
}
`;

exports[`large test: exact mode applies messages with multiple affected packages 1`] = `
Object {
  "/changelog/1.9.0": Object {
    "operation": "add",
    "value": Object {
      "changes": Object {
        "one": Object {
          "determinedIncrementLevel": "minor",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "",
              "subject": "waz",
              "type": "feat",
            },
          ],
          "version": "1.3.0",
        },
        "three": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "two",
                "three",
              ],
              "body": "
",
              "subject": "tar",
              "type": "fix",
            },
          ],
          "version": "3.4.6",
        },
        "two": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "two",
                "three",
              ],
              "body": "
",
              "subject": "tar",
              "type": "fix",
            },
          ],
          "version": "2.3.5",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "/mainPackage/rlsr/determinedIncrementLevel": Object {
    "operation": "add",
    "value": 1,
  },
  "/mainPackage/rlsr/previouslyUnreleased": Object {
    "operation": "add",
    "value": Array [],
  },
  "/mainPackage/version": Object {
    "operation": "update",
    "value": "1.9.0",
  },
  "/messages/0": Object {
    "operation": "delete",
  },
  "/messages/1": Object {
    "operation": "delete",
  },
  "/packages/one/version": Object {
    "operation": "update",
    "value": "1.3.0",
  },
  "/packages/three/version": Object {
    "operation": "update",
    "value": "3.4.6",
  },
  "/packages/two/version": Object {
    "operation": "update",
    "value": "2.3.5",
  },
}
`;

exports[`large test: exact mode applies minor message 1`] = `
Object {
  "/changelog/1.9.0": Object {
    "operation": "add",
    "value": Object {
      "changes": Object {
        "four": Object {
          "determinedIncrementLevel": "minor",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "",
              "subject": "bar",
              "type": "feat",
            },
          ],
          "version": "4.6.0",
        },
        "three": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "",
              "source": undefined,
              "subject": "feat: bar",
              "type": "dependency update",
            },
          ],
          "version": "3.4.6",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "/mainPackage/rlsr/determinedIncrementLevel": Object {
    "operation": "add",
    "value": 1,
  },
  "/mainPackage/rlsr/previouslyUnreleased": Object {
    "operation": "add",
    "value": Array [],
  },
  "/mainPackage/version": Object {
    "operation": "update",
    "value": "1.9.0",
  },
  "/messages/0": Object {
    "operation": "delete",
  },
  "/packages/four/version": Object {
    "operation": "update",
    "value": "4.6.0",
  },
  "/packages/three/dependencies/four": Object {
    "operation": "update",
    "value": "4.0.0 - 4",
  },
  "/packages/three/version": Object {
    "operation": "update",
    "value": "3.4.6",
  },
}
`;

exports[`large test: exact mode applies minor then major 1`] = `
Object {
  "/changelog/1.9.0": Object {
    "operation": "add",
    "value": Object {
      "changes": Object {
        "one": Object {
          "determinedIncrementLevel": "minor",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "",
              "subject": "raz",
              "type": "feat",
            },
          ],
          "version": "1.3.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "/mainPackage/rlsr/determinedIncrementLevel": Object {
    "operation": "add",
    "value": 1,
  },
  "/mainPackage/rlsr/previouslyUnreleased": Object {
    "operation": "add",
    "value": Array [],
  },
  "/mainPackage/version": Object {
    "operation": "update",
    "value": "1.9.0",
  },
  "/messages/0": Object {
    "operation": "delete",
  },
  "/packages/one/version": Object {
    "operation": "update",
    "value": "1.3.0",
  },
}
`;

exports[`large test: exact mode applies minor then major 2`] = `
Object {
  "/mainPackage/rlsr/determinedIncrementLevel": Object {
    "operation": "update",
    "value": -1,
  },
  "/messages/0": Object {
    "operation": "delete",
  },
  "/packages": Object {
    "operation": "update",
    "value": Object {
      "0": Object {
        "dependencies": Object {
          "five": "^4.5.0",
          "four": "^4.5.0",
        },
        "name": "one",
        "version": "1.3.0",
      },
      "1": Object {
        "dependencies": Object {
          "three": "2.3.4 - 3",
        },
        "name": "two",
        "version": "2.3.4",
      },
      "2": Object {
        "dependencies": Object {
          "four": "4.0.0 - 4.5.6",
        },
        "name": "three",
        "version": "3.4.5",
      },
      "3": Object {
        "name": "four",
        "version": "4.5.6",
      },
      "4": Object {
        "name": "five",
        "version": "4.5.6",
      },
    },
  },
}
`;

exports[`large test: exact mode applies patch message 1`] = `
Object {
  "/changelog/1.8.6": Object {
    "operation": "add",
    "value": Object {
      "changes": Object {
        "four": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "",
              "subject": "foo",
              "type": "fix",
            },
          ],
          "version": "4.5.7",
        },
        "three": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "",
              "source": undefined,
              "subject": "fix: foo",
              "type": "dependency update",
            },
          ],
          "version": "3.4.6",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "/mainPackage/rlsr/determinedIncrementLevel": Object {
    "operation": "add",
    "value": 0,
  },
  "/mainPackage/rlsr/previouslyUnreleased": Object {
    "operation": "add",
    "value": Array [],
  },
  "/mainPackage/version": Object {
    "operation": "update",
    "value": "1.8.6",
  },
  "/messages/0": Object {
    "operation": "delete",
  },
  "/packages/four/version": Object {
    "operation": "update",
    "value": "4.5.7",
  },
  "/packages/three/dependencies/four": Object {
    "operation": "update",
    "value": "4.0.0 - 4",
  },
  "/packages/three/version": Object {
    "operation": "update",
    "value": "3.4.6",
  },
}
`;

exports[`large test: exact mode applies several messages with one breaking change 1`] = `
Object {
  "/changelog/5.0.0": Object {
    "operation": "add",
    "value": Object {
      "changes": Object {
        "five": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "four",
              ],
              "subject": "dependency update: fix: taz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
        "four": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "subject": "taz",
              "type": "fix",
            },
          ],
          "version": "5.0.0",
        },
        "one": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "",
              "subject": "raz",
              "type": "fix",
            },
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "four",
              ],
              "subject": "dependency update: fix: taz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
        "three": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "three",
              ],
              "body": "",
              "subject": "roo",
              "type": "fix",
            },
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "four",
              ],
              "subject": "dependency update: fix: taz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
        "two": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "source": Array [
                "four",
              ],
              "subject": "dependency update: fix: taz",
              "type": "dependency update",
            },
          ],
          "version": "5.0.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "/mainPackage/rlsr/determinedIncrementLevel": Object {
    "operation": "add",
    "value": 2,
  },
  "/mainPackage/rlsr/previouslyUnreleased": Object {
    "operation": "add",
    "value": Array [],
  },
  "/mainPackage/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/messages/0": Object {
    "operation": "delete",
  },
  "/messages/1": Object {
    "operation": "delete",
  },
  "/messages/2": Object {
    "operation": "delete",
  },
  "/packages/five/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/four/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/one/dependencies/five": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/one/dependencies/four": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/one/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/three/dependencies/four": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/three/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
  "/packages/two/dependencies/three": Object {
    "operation": "update",
    "value": "^5.0.0",
  },
  "/packages/two/version": Object {
    "operation": "update",
    "value": "5.0.0",
  },
}
`;
